### DISCUSSION TITLE
Implement ToLua for a particular HashMap type

### DISCUSSION THREAD

**Discussion by RageCage64** [mlua-rs/mlua#271] [Q&A] [answered]
Hello! Thanks for all the work on this library.

What I'm trying to accomplish is for a user to specify a table in a file that will get deserialized using `serde` and passed to a Lua script as an argument. My attempt to accomplish this is by starting with this enum:

```rust
#[derive(Deserialize, Debug)]
pub enum RecordValidType {
    String(String),
    Number(f64),
    Table(HashMap<String, RecordValidType>),
}
```
And it's being used in another deserializing struct like this:
```rust
#[derive(Deserialize, Debug)]
pub struct LuaFnInput {
    pub tag: String,
    pub timestamp: String,
    pub record: HashMap<String, RecordValidType>,
}
```
I tried to implement the `ToLua` trait on the enum type like this:
```rust
// error handling removed for brevity, just assume everything is perfect hooray
impl<'lua> ToLua<'lua> for RecordValidType {
    fn to_lua(self, lua: &'lua mlua::Lua) -> mlua::Result<Value<'lua>> {
        match self {
            String => return Ok(self.to_lua(lua).unwrap()),
            Number => return Ok(self.to_lua(lua).unwrap()),
            Table => {
                // In a theoretical world where this works
                let record: HashMap<String, RecordValidType> = HashMap::from(self);
                let table = lua.create_table().unwrap();
                for row in record.iter() {
                    table.set(*row.0, *row.1);
                }
                return Ok(table);
            }
        }
    }
}
```
However this is pretty clearly wrong, and I'm not sure how to pull it off. Is this something that is possible? Am I completely off-base and there's a better way to accomplish this? Thank you in advance!

**Comment by khvzak** [comment, marked as answer]
Probably you don't even need `ToLua` implementation for `RecordValidType`.
mlua (with `serialize` feature flag) support integration with serde. If you add `derive(Serialize)` to `RecordValidType` then should be able to automatically convert it to Lua.
Eg:
```rust
let lua_val = lua.to_value(&RecordValidType::Number(1.0))?;
```

You just need to import `LuaSerdeExt` trait.

**Reply by RageCage64** [reply]
Thank you for the help! Trying to force `to_lua` when I could have just been using `lua.to_value` with the serde extension the whole time was definitely the way to go. I didn't actually know about that function and it didn't pop out to me reading the documentation; hopefully anyone else with the same problem will happen by this issue to for this answer.

Thank you for your help!
